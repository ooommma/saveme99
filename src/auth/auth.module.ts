import { Module, forwardRef } from '@nestjs/common';
import { AuthService } from './auth.service';
import { AuthController } from './auth.controller';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Users } from '../user/entities/users.entity';
import { ConfigService } from '@nestjs/config';
import { JwtModule } from '@nestjs/jwt';
import { PassportModule } from '@nestjs/passport';
import { JwtStrategy } from './jwt.startegy';
import { UserModule } from 'src/user/user.module';
import { AwsModule } from 'src/aws/aws.module';
import { UtilsModule } from 'src/utils/utils.module';
import { UserService } from 'src/user/user.service';
@Module({
  imports: [
    PassportModule.register({ defaultStrategy: 'jwt', session: false }),

    JwtModule.registerAsync({
      useFactory: (config: ConfigService) => ({
        secret: config.get('JWT_SECRET'),
        signOptions: {
          expiresIn: 60 * 60 * 5,
        },
      }),
      inject: [ConfigService],
    }),
    TypeOrmModule.forFeature([Users]),
    forwardRef(() => UserModule),
    AwsModule,
    UtilsModule,
  ],
  providers: [AuthService, JwtStrategy],
  controllers: [AuthController],
  exports: [PassportModule, AuthService],
})
export class AuthModule {}
