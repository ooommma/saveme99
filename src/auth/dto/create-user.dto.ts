import { IsEmail, IsString, MaxLength, MinLength } from 'class-validator';

export class CreateUserDto {
  @IsString()
  @IsEmail()
  email: string;

  @IsString()
  @MaxLength(15)
  @MinLength(4)
  password: string;

  profileImg?: string;

  @IsString()
  @MaxLength(10)
  @MinLength(2)
  name: string;
}
