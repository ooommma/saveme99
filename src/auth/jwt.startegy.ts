import {
  Injectable,
  Req,
  Request,
  UnauthorizedException,
} from '@nestjs/common';
import { PassportStrategy } from '@nestjs/passport';
import { InjectRepository } from '@nestjs/typeorm';
import { ExtractJwt, Strategy } from 'passport-jwt';

import { Repository } from 'typeorm';
import { Request as RequestType } from 'express';
import { ConfigService } from '@nestjs/config';
import { Users } from '../user/entities/users.entity';

@Injectable()
export class JwtStrategy extends PassportStrategy(Strategy) {
  constructor(
    private configService: ConfigService,
    @InjectRepository(Users)
    private userRepository: Repository<Users>,
  ) {
    super({
      jwtFromRequest: ExtractJwt.fromExtractors([
        JwtStrategy.extractJWT,
        ExtractJwt.fromAuthHeaderAsBearerToken(),
      ]),
      ignoreExpiration: false,
      secretOrKey: configService.get('JWT_SECRET'),
    });
  }

  private static extractJWT(req: RequestType): string | null {
    if (
      req.cookies &&
      'authorization' in req.cookies &&
      req.cookies.authorization.length > 0
    ) {
      return req.cookies.authorization.split(' ')[1];
    }
    return null;
  }

  async validate(payload) {
    const { userId } = payload;
    const user: Users = await this.userRepository.findOne({ where: { userId } });

    if (!user) {
      throw new UnauthorizedException();
    }

    return user;
  }
}
