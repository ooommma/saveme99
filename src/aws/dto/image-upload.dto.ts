import { ApiProperty } from '@nestjs/swagger';
import { IsNotEmpty } from 'class-validator';

export class ImageUploadDto {
  @ApiProperty({ type: 'string', format: 'binary' })
  file: any;
}
