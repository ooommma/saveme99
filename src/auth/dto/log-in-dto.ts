import { PartialType, PickType } from '@nestjs/mapped-types';
import { CreateUserDto } from './create-user.dto';

export class UserLoginDto extends PickType(CreateUserDto, ['email', 'password']) {}
