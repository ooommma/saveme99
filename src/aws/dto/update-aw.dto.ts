import { PartialType } from '@nestjs/mapped-types';
import { CreateAwDto } from './create-aw.dto';

export class UpdateAwDto extends PartialType(CreateAwDto) {}
