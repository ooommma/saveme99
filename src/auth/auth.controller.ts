import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  Res,
  UnauthorizedException,
  UseGuards,
} from '@nestjs/common';
import { AuthService } from './auth.service';
import { CreateUserDto } from './dto/create-user.dto';
import { UserLoginDto } from './dto/log-in-dto';
import { Response } from 'express';
import { AuthGuard } from '@nestjs/passport';
import { GetUser } from './decorator/get-user.decorator';
import { User } from './entities/auth.entity';

@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Post('/register')
  async signUp(@Body() createAuthDto: CreateUserDto) {
    return await this.authService.createUser(createAuthDto);
  }

  @Post('/login')
  async logIn(@Body() UserLoginDto: UserLoginDto, @Res() res: Response) {
    const { accessToken, refreshToken } = await this.authService.logIn(
      UserLoginDto,
      res,
    );
    if (!accessToken || !refreshToken) {
      throw new UnauthorizedException();
    }
    const BearerAccessToken = `Bearer ${accessToken}`;
    const BearerRefreshToken = `Bearer ${refreshToken}`;
    res.cookie('authorization', BearerAccessToken, {
      maxAge: 1000 * 60 * 60,
      httpOnly: true,
      sameSite: true,
    });
    res.cookie('refreshToken', BearerRefreshToken, {
      maxAge: 1000 * 60 * 60 * 24 * 7,
      httpOnly: true,
      sameSite: true,
    });
    res.status(200).json({ message: '로그인 성공' });
  }

  @Get()
  findAll() {
    return this.authService.findAll();
  }
  @Get('profile')
  @UseGuards(AuthGuard('jwt'))
  findOne(@GetUser() user: User) {
    return user;
  }

  // @Patch(':id')
  // update(@Param('id') id: string, @Body() updateAuthDto: UpdateAuthDto) {
  //   return this.authService.update(+id, updateAuthDto);
  // }

  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.authService.remove(+id);
  }
}