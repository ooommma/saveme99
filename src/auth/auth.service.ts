import {
  ConflictException,
  Injectable,
  NotFoundException,
  UnauthorizedException,
} from '@nestjs/common';
import { CreateUserDto } from './dto/create-user.dto';
import { InjectRepository } from '@nestjs/typeorm';
import { User } from './entities/auth.entity';
import { Repository } from 'typeorm';
import { UpdateUserDto } from './dto/update-auth.dto';
import { hash, compare } from 'bcrypt';
import { UserLoginDto } from './dto/log-in-dto';
import { Response } from 'express';
import jwt from 'jsonwebtoken';
import { JwtService } from '@nestjs/jwt';

@Injectable()
export class AuthService {
  constructor(
    @InjectRepository(User)
    private userRepository: Repository<User>,
    private jwtService: JwtService,
  ) {}
  async createUser(createUserDto: CreateUserDto) {
    const { email, password } = createUserDto;
    const user = await this.userRepository.findOneBy({ email });
    if (user) {
      throw new ConflictException('이미 회원가입한 이메일입니다.');
    }
    const hashedPassword = await hash(password, 10);
    const userDao = await this.userRepository.save({
      ...createUserDto,
      password: hashedPassword,
    });
    return userDao;
  }

  async logIn(userLoginDto: UserLoginDto, res: Response) {
    const { email, password } = userLoginDto;
    const user = await this.findUserByEmailWithPassword(email);
    //유저가 있다면 비밀번호 검증
    console.log(user);
    if (!compare(password, user.password)) {
      throw new UnauthorizedException('비밀번호를 다시 확인해주세요.');
    }
    const payload: object = { userId: user.userId };
    console.log(payload);
    const accessToken = this.jwtService.sign(payload, {
      expiresIn: 1000 * 60 * 60,
    });
    const refreshToken = this.jwtService.sign(payload, {
      expiresIn: 1000 * 60 * 60 * 24 * 7,
    });
    return { accessToken, refreshToken };
  }

  findAll() {
    return `This action returns all auth`;
  }

  async findUserByEmail(email: string): Promise<User> {
    const user = await this.userRepository.findOneBy({ email });
    if (!user)
      throw new NotFoundException(
        `${email}에 해당하는 유저를 찾을 수 없습니다.`,
      );
    return user;
  }

  async findUserByEmailWithPassword(email: string): Promise<User> {
    const user = await this.userRepository.findOne({
      where: { email },
      select: ['email', 'password', 'userId'],
    });
    if (!user)
      throw new NotFoundException(
        `${email}에 해당하는 유저를 찾을 수 없습니다.`,
      );
    return user;
  }

  update(id: number, updateAuthDto: UpdateUserDto) {
    return `This action updates a #${id} auth`;
  }

  remove(id: number) {
    return `This action removes a #${id} auth`;
  }
}