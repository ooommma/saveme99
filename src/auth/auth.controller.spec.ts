import { Test, TestingModule } from '@nestjs/testing';
import { AuthController } from './auth.controller';
import { AuthService } from './auth.service';
import { after } from 'node:test';
import { Readable } from 'stream';
import { Response } from 'express';
import { ForbiddenException, UnauthorizedException } from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { Users } from 'src/user/entities/users.entity';
import { CreateUserDto } from 'src/user/dto/create-user.dto';
import { validate } from 'class-validator';
import { error } from 'console';
import { UserLoginDto } from './dto/log-in-dto';

const mockAuthService = {
  createUser: jest.fn(),
  logIn: jest.fn(),
};

const mockResponse = (): Response => {
  const res: Partial<Response> = {};
  res.status = jest.fn().mockReturnValue(res);
  res.json = jest.fn().mockReturnValue(res);
  res.send = jest.fn().mockReturnValue(res);
  res.cookie = jest.fn().mockReturnValue(res);
  return res as Response;
};

interface IUser {
  email: string;
  password: string;
  name: string;
}

const createUserDto: IUser = {
  email: 'popcon940620@gmail.com',
  password: '123456',
  name: '김라임',
};

const mockFile: Express.Multer.File = {
  fieldname: 'file',
  originalname: 'testfile.txt',
  encoding: '7bit',
  mimetype: 'text/plain',
  size: 128,
  destination: './upload',
  filename: 'testfile.txt',
  path: './upload/testfile.txt',
  buffer: Buffer.from('Hello World'),
  stream: Readable.from(Buffer.from('Hello World')),
};
const mockUser: Users = {
  userId: 1,
  email: 'popcon940620@gmail.com',
  password: '42412412',
  name: '김라임',
  profileImg: null,
  createdAt: new Date(),
  updatedAt: new Date(),
};

let canActivateVal = true;

const mockAuthGuard = { canActivate: jest.fn(() => canActivateVal) };

describe('AuthController with AuthGuard true', () => {
  let authController: AuthController;
  let authService: AuthService;
  let res: Partial<Response>;

  beforeEach(async () => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.restoreAllMocks();

    res = mockResponse();

    const module: TestingModule = await Test.createTestingModule({
      controllers: [AuthController],
      providers: [
        {
          provide: AuthService,
          useValue: mockAuthService,
        },
      ],
    })
      .overrideProvider(AuthGuard('jwt'))
      .useValue(mockAuthGuard)
      .compile();

    authController = module.get<AuthController>(AuthController);
    authService = module.get<AuthService>(AuthService);
  });

  afterAll(async () => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.restoreAllMocks();
  });

  it('should be defined', () => {
    expect(authController).toBeDefined();
    expect(authService).toBeDefined();
  });

  describe('createUser Method', () => {
    it('success createUser', async () => {
      const mockUser = 'im User';

      mockAuthService.createUser.mockResolvedValue(mockUser);

      const user = await authController.signUp(createUserDto, mockFile);

      expect(mockAuthService.createUser).toHaveBeenCalledWith(createUserDto, mockFile);
      expect(mockAuthService.createUser).toHaveBeenCalledTimes(1);
      expect(user).toEqual(mockUser);
    });
    it('signUp should be failed by validate', async () => {
      const dto = new CreateUserDto();
      dto.name = '김';
      dto.email = 'popcon940620';
      dto.password = '123';
      const errors = await validate(dto);
      expect(errors.length).toBeGreaterThan(0);
      expect(errors).toMatchSnapshot();
    });
  });

  describe('login Method', () => {
    it('success Login', async () => {
      const res = mockResponse();

      const mockTokens = {
        accessToken: 'im access token',
        refreshToken: 'im refresh token',
      };
      const mockBearerRefreshToken = 'Bearer im refresh token';
      const mockUserLoginDto = {
        email: 'popcon940620@gmail.com',
        password: '123456',
      };

      mockAuthService.logIn.mockResolvedValue(mockTokens);

      await authController.logIn(mockUserLoginDto, res);

      expect(mockAuthService.logIn).toHaveBeenCalledTimes(1);
      expect(mockAuthService.logIn).toHaveBeenCalledWith(mockUserLoginDto, res);
      expect(res.cookie).toHaveBeenCalledTimes(2);
      expect(res.cookie).toHaveBeenLastCalledWith('refreshToken', mockBearerRefreshToken, {
        maxAge: 1000 * 60 * 60 * 24 * 7,
        httpOnly: true,
        sameSite: true,
      });
      expect(res.status).toHaveBeenCalledTimes(1);
      expect(res.status).toHaveBeenCalledWith(200);
      expect(res.json).toHaveBeenCalledTimes(1);
      expect(res.json).toHaveBeenCalledWith({ message: '로그인 성공' });
    });
    it('login method should be throw Unauthorization Exception', async () => {
      const res = mockResponse();

      const mockTokens = {
        accessToken: 'im access token',
      };

      const mockUserLoginDto = {
        email: 'popcon940620@gmail.com',
        password: '123456',
      };
      mockAuthService.logIn.mockResolvedValue(mockTokens);

      await expect(authController.logIn(mockUserLoginDto, res)).rejects.toThrow(UnauthorizedException);

      expect(mockAuthService.logIn).toHaveBeenCalledTimes(1);
      expect(mockAuthService.logIn).toHaveBeenCalledWith(mockUserLoginDto, res);
    });
    it('UserLoginDto should be failed by validate', async () => {
      const dto = new UserLoginDto();
      dto.email = 'popcon940620';
      dto.password = '123';

      const errors = await validate(dto);
      expect(errors.length).toBeGreaterThan(0);
      expect(errors).toMatchSnapshot();
    });
  });

  describe('profile method', () => {
    it('success profile method', async () => {
      expect(authController.getProfile(mockUser)).toEqual(mockUser);
    });
  });
});
