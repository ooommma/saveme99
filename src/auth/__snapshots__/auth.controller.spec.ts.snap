// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthController with AuthGuard true createUser Method signUp should be failed by validate 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isEmail": "email must be an email",
    },
    "property": "email",
    "target": CreateUserDto {
      "email": "popcon940620",
      "name": "김",
      "password": "123",
    },
    "value": "popcon940620",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "minLength": "password must be longer than or equal to 4 characters",
    },
    "property": "password",
    "target": CreateUserDto {
      "email": "popcon940620",
      "name": "김",
      "password": "123",
    },
    "value": "123",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "minLength": "name must be longer than or equal to 2 characters",
    },
    "property": "name",
    "target": CreateUserDto {
      "email": "popcon940620",
      "name": "김",
      "password": "123",
    },
    "value": "김",
  },
]
`;

exports[`AuthController with AuthGuard true login Method UserLoginDto should be failed by validate 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isEmail": "email must be an email",
    },
    "property": "email",
    "target": UserLoginDto {
      "email": "popcon940620",
      "password": "123",
    },
    "value": "popcon940620",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "minLength": "password must be longer than or equal to 4 characters",
    },
    "property": "password",
    "target": UserLoginDto {
      "email": "popcon940620",
      "password": "123",
    },
    "value": "123",
  },
]
`;
